runtime/oti/VMHelpers.hpp

```diff
@@ -2043,8 +2043,23 @@ class VM_VMHelpers
 	threadInterruptImpl(J9VMThread *currentThread, j9object_t targetObject)
 	{
 		J9VMThread *targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, targetObject);
-		if (J9VMJAVALANGTHREAD_STARTED(currentThread, targetObject) && (NULL != targetThread)) {
-			void (*sidecarInterruptFunction)(J9VMThread*) = currentThread->javaVM->sidecarInterruptFunction;
+		J9JavaVM *vm = currentThread->javaVM;
+#if JAVA_SPEC_VERSION >= 19
+		/* Check if the mounted thread is suspended. */
+		U_32 isSuspended = J9OBJECT_U32_LOAD(currentThread, targetThread->threadObject, vm->isSuspendedInternalOffset);
+#endif /* JAVA_SPEC_VERSION >= 19 */
+		if ((NULL != targetThread)
+		&& J9VMJAVALANGTHREAD_STARTED(currentThread, targetObject)
+#if JAVA_SPEC_VERSION >= 19
+		/* Thread.deadInterrupt is Thread.interrupted in OJDK's Thread implementation.
+		 * In JDK19+, OJDK's Thread implementation is used. If the mounted thread is
+		 * suspended, only set Thread.interrupted to TRUE and do not wake/interrupt
+		 * the thread.
+		 */
+		&& !isSuspended
```

Comment:
A few questions: 
Does this mean a `suspended` thread can't be interrupted?
Is this a RI behaviour as well?
Is a similar check required at `threadIsInterruptedImpl()`?
https://github.com/eclipse-openj9/openj9/blob/4be0f692337ff3ee37cd0fa8bdafe321492620c8/runtime/oti/VMHelpers.hpp#L1413-L1421

Reply:
> Does this mean a suspended thread can't be interrupted?

A suspended thread shouldn't run. When it is interrupted, only the `Thread.interrupted` field is set to signify that the thread has been interrupted.

> Is this a RI behaviour as well?

Yes, a test has been added where the RI behaves in this manner: [libGetThreadStateMountedTest.cpp#L168-L177](https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/cdc12a56665796db5270385de01dc6a30804ba03/test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp#L168-L177). 

> Is a similar check required at threadIsInterruptedImpl()?

In the OJDK Thread impl, we can just rely on the `Thread.interrupted` field to determine if the thread is interrupted. But, there is CRIU specific code in [Thread.isInterrupted()](https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/cdc12a56665796db5270385de01dc6a30804ba03/src/java.base/share/classes/java/lang/Thread.java#L1757-L1762) which invokes `threadIsInterruptedImpl()`. So, we will need a similar check in `threadIsInterruptedImpl()`.

Reply:
@JasonFengJ9 I have updated the PR. Can you please take another look?

Reply:
A further question:
> A suspended thread shouldn't run. When it is interrupted, only the Thread.interrupted field is set to signify that the thread has been interrupted.

Considering a mounted thread is blocked in an invocation of the `Object.wait()`, and it is suspended, assuming this thread state is `Thread.State.WAITING`, does its caller receive an `InterruptedException` when it is interrupted?

Reply:
> Considering a mounted thread is blocked in an invocation of the Object.wait(), and it is suspended, assuming this thread state is Thread.State.WAITING, does its caller receive an InterruptedException when it is interrupted?

The behaviour between a suspended thread and it receiving an interrupt is not properly defined in the Java doc and JVMTI spec. In the same test, they have added a case for the WAITING scenario: [GetThreadStateMountedTest.java#L102-L151](https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/cdc12a56665796db5270385de01dc6a30804ba03/test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java#L102-L151). In a comment, it is mentioned that `for safety, Object.wait should throw InterruptedException`. RI seems to throw the InterruptedException after the thread is resumed, not while it is suspended; in some cases, I didn't see my print statement from the InterruptedException catch block; it seems like the RI is inconsistent in throwing the exception. 

My current changes won't throw an `InterruptedException`. We probably need to throw the exception once the thread is resumed.

Reply:
@JasonFengJ9 What are your thoughts on handling this issue?

https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/Thread.html#interrupt()
> If this thread is blocked in an invocation of the [wait()](https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/Object.html#wait()), [wait(long)](https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/Object.html#wait(long)), or [wait(long, int)](https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/Object.html#wait(long,int)) methods of the [Object](https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/Object.html) class, or of the [join()](https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/Thread.html#join()), [join(long)](https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/Thread.html#join(long)), [join(long, int)](https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/Thread.html#join(long,int)), [sleep(long)](https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/Thread.html#sleep(long)), or [sleep(long, int)](https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/Thread.html#sleep(long,int)) methods of this class, then its interrupt status will be cleared and it will receive an [InterruptedException](https://download.java.net/java/early_access/jdk22/docs/api/java.base/java/lang/InterruptedException.html).

- **Option 1:** We can read the `Thread.interrupt` field, set `InterruptedException`, and clear the interrupt in [threadhelp.cpp::monitorWaitImpl](https://github.com/eclipse-openj9/openj9/blob/4be0f692337ff3ee37cd0fa8bdafe321492620c8/runtime/vm/threadhelp.cpp#L109) and [threadhelp.cpp::threadSleepImpl](https://github.com/eclipse-openj9/openj9/blob/4be0f692337ff3ee37cd0fa8bdafe321492620c8/runtime/vm/threadhelp.cpp#L175).
- **Option 2:** We can read the `Thread.interrupt` field, set `InterruptedException`, and clear the interrupt in [JVMTI ResumeThread](https://github.com/eclipse-openj9/openj9/blob/4be0f692337ff3ee37cd0fa8bdafe321492620c8/runtime/jvmti/jvmtiThread.cpp#L1172).

Reply:
@JasonFengJ9 See the [new changes](https://github.com/eclipse-openj9/openj9/compare/fdd8574b404e58cbf6188de7b03056b1b7a966d0..806dfa59c82ee19d4ed968bdc3e30ccf671e73ad). They make sure that the thread throws `InterruptedException` once it is woken up after being resumed. Thus, matching the RI.

Reply:
[VirtualThread.sleep()](https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/cdc12a56665796db5270385de01dc6a30804ba03/src/java.base/share/classes/java/lang/VirtualThread.java#L797-L798) seems to park the thread instead, hence `threadSleepImpl()` change only affects those non-virtual threads.
To be on the safe side, probably should verify RI throwing `InterruptedException` at [waiting()](https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/cdc12a56665796db5270385de01dc6a30804ba03/test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java#L128-L130) and [sleeping()](https://github.com/ibmruntimes/openj9-openjdk-jdk22/blob/cdc12a56665796db5270385de01dc6a30804ba03/test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java#L170-L172)

Reply:
RI never throws an `InterruptedException` for `waiting(timeout)` and `sleeping`. It intermittently throws an `InterruptedException` for `waiting` indefinitely (without a `timeout`), but most of the time it didn't throw an `InterruptedException` even for this case. The test is missing asserts to validate `InterruptedException` because the RI doesn't guarantee to throw an `InterruptedException`.

My last push to throw an `InterruptedException` in OpenJ9 has side-effects so I have reverted. It introduces a race condition where a notified thread might falsely assume that it has been interrupted. Addressing the race condition will be challenging. Due to RI's inconsistent behaviour, we should separately handle this change outside the scope of this PR. I will open another issue to track it.

Reply:
Thanks the investigation, no further questions.